### PR TITLE
Feature/cvsb 12215

### PR DESCRIPTION
**Pull request by oalexandru**

### CODE REVIEW

File: src/services/TestResultsService.ts [hunk 1]
```diff
@@ -309,14 +307,21 @@ export class TestResultsService {
...
                         console.log(`Setting expiryDate: ${testType.testExpiryDate}`);
                       } else {
                         // less than 2 months then set expiryDate 1 year after the Registration/First Use Anniversary date
                         testType.testExpiryDate = dateFns.addYears(anvDateForCompare, 1).toISOString();
                         console.log(`Setting expiryDate as 1yr from RegDate: ${testType.testExpiryDate}`);
                       }
+                    } else if (this.isAnnualTestRetestTestType(testType) && dateFns.isEqual(mostRecentExpiryDateOnAllTestTypesBySystemNumber, new Date(1970, 1, 1))) {
+                        const registrationFirstUseAnniversaryDate = dateFns.addYears(dateFns.lastDayOfMonth(regOrFirstUseDate!), 1);
+                        if (!regOrFirstUseDate || dateFns.isBefore(dateFns.addMonths(new Date(), 2), registrationFirstUseAnniversaryDate) || dateFns.isEqual(dateFns.addMonths(new Date(), 2), registrationFirstUseAnniversaryDate)) {
+                          testType.testExpiryDate = dateFns.addHours(this.lastDayOfMonthInNextYear(new Date()), 12).toISOString();
```
[the remark this answers is not in the document]
Comment (KHodderDVSA) [reply src/services/TestResultsService.ts]: I can no longe replicate the previous issue with v1.30.1 of date-fns. Might be worth updating and seeing if you still need it

File: src/services/TestResultsService.ts [hunk 2]
```diff
@@ -309,14 +307,21 @@ export class TestResultsService {
...
                         testType.testExpiryDate = this.lastDayOfMonthInNextYear(new Date()).toISOString();
                         console.log(`Setting expiryDate: ${testType.testExpiryDate}`);
                       } else {
                         // less than 2 months then set expiryDate 1 year after the Registration/First Use Anniversary date
                         testType.testExpiryDate = dateFns.addYears(anvDateForCompare, 1).toISOString();
                         console.log(`Setting expiryDate as 1yr from RegDate: ${testType.testExpiryDate}`);
                       }
+                    } else if (this.isAnnualTestRetestTestType(testType) && dateFns.isEqual(mostRecentExpiryDateOnAllTestTypesBySystemNumber, new Date(1970, 1, 1))) {
+                        const registrationFirstUseAnniversaryDate = dateFns.addYears(dateFns.lastDayOfMonth(regOrFirstUseDate!), 1);
+                        if (!regOrFirstUseDate || dateFns.isBefore(dateFns.addMonths(new Date(), 2), registrationFirstUseAnniversaryDate) || dateFns.isEqual(dateFns.addMonths(new Date(), 2), registrationFirstUseAnniversaryDate)) {
```
Comment (KHodderDVSA) [inline src/services/TestResultsService.ts]: I don't understand the third condition on this. It looks the same as the second - I would just compare to "the end of the registration/firstUse anniversary date day" - there's an dateFns.endOfDay function for that.

File: tests/unit/generateExpiryDate.unitTest.ts [hunk 3]
```diff
@@ -243,6 +243,75 @@ describe("TestResultsService calling generateExpiryDate", () => {
...
+
+                    const expectedExpiryDate = dateFns.setHours(dateFns.lastDayOfMonth(dateFns.addYears(new Date(), 1)), 12);
+                    return testResultsService.generateExpiryDate(hgvTestResult)
+                        .then((hgvTestResultWithExpiryDate: any) => {
+                            expect((hgvTestResultWithExpiryDate.testTypes[0].testExpiryDate).split("T")[0]).toEqual(expectedExpiryDate.toISOString().split("T")[0]);
+                        });
+                });
+            });
+
+            context("when there is an Annual Test Type with no existing expiryDate and testDate is less than 2 months before Registration Anniversary date.", () => {
```
Comment (KHodderDVSA) [inline tests/unit/generateExpiryDate.unitTest.ts]: no test for "same day"

Comment (oalexandru) [reply tests/unit/generateExpiryDate.unitTest.ts]: by "same day" I believe you mean "testdate is exactly 2 months before registration anniversary date". Fixed

File: src/services/TestResultsService.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (oalexandru) [reply src/services/TestResultsService.ts]: 1.30.1 is the version I'm trying with now. Tried with 2.0.0 with no success. I adopted your other suggestion though. Thanks

File: src/services/TestResultsService.ts (hunk 2 above)
Comment (oalexandru) [reply src/services/TestResultsService.ts]: it's for further enforcing the case: "test date exactly 2 months before the registration anniversary/ first use anniversary". Apparently the 2nd condition is enough, but given the latest issues we had with dateFns I considered the 3rd wouldn't hurt.

Comment (oalexandru) [reply src/services/TestResultsService.ts]: fixed with your suggestion

**Pull request merged**